src/tests/authentication/test_token.py

```diff
@@ -0,0 +1,222 @@
+            payload,
+        )
+
+    def get_token(self, response=None, **kwargs):
+        if response is None:
+            response = self.obtain_token(**kwargs)
+
+        return json.loads(response.content)['token']
+
+    def assertValidJson(self, content: str):
+        try:
+            return json.loads(content)
+        except json.JSONDecodeError:
+            self.fail('Failed to parse JSON')
+
+    def api_call(self, path: str, payload: dict = None, token: str = None, method: str = 'get',
+                 decode_json: bool = True):
+        if payload is None:
+            payload = {}
+
+        headers = {}
+        if token is not None:
+            headers['HTTP_AUTHORIZATION'] = f'Token {token}'
+
+        if method == 'get':
+            response = self.client.get(
+                path,
+                payload,
+                **headers,
+            )
+        elif method == 'post':
+            response = self.client.post(
+                path,
+                payload,
+                **headers,
+            )
+        elif method == 'delete':
+            response = self.client.delete(
+                path,
+                payload,
+                **headers,
+            )
+        else:
+            raise ValueError('method mus be either get or post')
+
+        if decode_json:
+            self.assertValidJson(response.content)
+            return response, json.loads(response.content)
+        else:
+            return response, response.content
+
+    def validate_token_against_api(self, token: Optional[str] = None, expected_status_code: Optional[int] = 200):
+        response, response_content = self.api_call('/api/authentication/token/', token=token)
+        if expected_status_code is not None:
+            self.assertEqual(response.status_code, expected_status_code)
+        return response, response_content
+
+    def test_obtain_token_json(self):
+        response = self.obtain_token()
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response['content-type'], 'application/json')
+        token = self.assertValidJson(response.content)
+        self.assertTrue('token' in token)
+
+    def test_token_bad_password(self):
+        response = self.obtain_token(password='')
+        self.assertEqual(response.status_code, 400)
+        self.assertValidJson(response.content)
+
+    def test_inactive_user_token(self):
+        user = create_user(
+            password=self.user_password,
+        )
+        token = self.get_token(
+            username=user.username,
+            password=self.user_password,
+        )
+        user.is_active = False
+        user.save()
+        self.validate_token_against_api(token, 401)
+
+    def test_token_expiry(self):
+        expiring = datetime.now() + timedelta(seconds=1)
+        token = self.get_token(expiring=expiring)
+        self.validate_token_against_api(token)
+        sleep(1.1)
```

Comment:
I'm not sure about sleeping in a unit test...
Do we have another way?

Reply:
I'm, also not that happy with that, but I found no other way. We need to ensure that the token is usable before it expired and non usable after it expired. In that case I would see one seconds not as the biggest hit.
If you have any other way, great. But modifying the expiry date afterwards would kinda defeat the point I think.

Reply:
Maybe we could mock the datetime with our own, manipulated time?

Reply:
Hm, Could you explain that a bit further?

Reply:
Implemented mocking in f20e5aa